README
## BridgeDB [![Build Status](https://travis-ci.org/isislovecruft/bridgedb.svg)](https://travis-ci.org/isislovecruft/bridgedb) [![Coverage Status](https://coveralls.io/repos/isislovecruft/bridgedb/badge.png)](https://coveralls.io/r/isislovecruft/bridgedb)

BridgeDB is a collection of backend servers used to distribute
[Tor Bridges](https://www.torproject.org/docs/bridges). It currently consists
of a webserver with [an HTTPS interface](https://bridges.torproject.org), an
email responder, and an SQLite database.

#### What are Tor Bridges?
[Tor Bridges](https://www.torproject.org/docs/bridges) are special Tor relays
which are not listed in the public relay directory. They are used to help
circumvent [censorship](https://ooni.torproject.org) by providing users with
connections to the public relays in the Tor network.

Tor Bridges are different from normal relays in another important way: they
can run what are called *Pluggable* *Transports*.

#### What's a Pluggable Transport?
A
[Pluggable Transport](https://www.torproject.org/docs/pluggable-transports.html.en)
is a program which is *pluggable* — meaning that it is meant to work with lots
of other anonymity and censorship circumvention software, not just Tor — and
is a *transport* — meaning that it transports your internet traffic, usually
in a way which makes it look different. For example,
[Obfsproxy](https://www.torproject.org/projects/obfsproxy.html.en) is a
Pluggable Transport which disguises your traffic by adding an obfuscating
layer of encryption.

#### So how do I use this?
Well, probably, you don't. But if you're looking for bridges, you can use
[the web interface](https://bridges.torproject.org) of the BridgeDB instance
deployed by the Tor Project, which has instructions on getting the Pluggable
Transports-capable Tor Browser Bundle, as well as instructions for getting
extra Bridges.


## Maintainer Setup

### Dependencies and installation
BridgeDB requires the following OS-level dependencies:

 - python>=2.7
 - python-dev
 - build-essential
 - libgpgme11
 - libgpgme11-dev
 - OpenSSL>=1.0.1e
 - [SQLite3](http://www.maxmind.com/app/python)
 - [MaxMind GeoIP](https://www.maxmind.com/en/geolocation_landing)
   - libgeoip-dev
   - geoip-database
 - [python-setuptools](https://pypi.python.org/pypi/setuptools)

As well as any Python dependencies in requirements.txt.

### Deploying BridgeDB

BridgeDB should work with or without a Python virtualenv.

 - Install Python 2.7, and other OS-level dependencies. On Debian, you can do:

        sudo apt-get install build-essential openssl python python-dev \
            python-setuptools sqlite3 libgpgme11 libgpgme11-dev libgeoip-dev \
            geoip-database

 - Install Pip 1.3.1 or later. Debian has this version, but if for some reason
   that or a newer version isn't available, the easiest way to install a newer
   Pip is to use the Pip development teams's [getpip
   script](https://raw.github.com/pypa/pip/master/contrib/get-pip.py):

        wget https://raw.github.com/pypa/pip/master/contrib/get-pip.py
        sudo python get-pip.py

 - (virtualenv installs only) Use Pip to install virtualenv and
   [virtualenvwrapper](https://virtualenvwrapper.readthedocs.org):

        sudo pip install --upgrade virtualenv virtualenvwrapper

 - (virtualenv installs only) Configure virtualenvwrapper and create a
   virtualenv for Bridgedb:

        WORKON_HOME=${HOME}/.virtualenvs
        export WORKON_HOME
        mkdir -p $WORKON_HOME
        source $(which virtualenvwrapper.sh)
        git clone https://git.torproject.org/bridgedb.git && cd bridgedb
        mkvirtualenv -a $PWD -r requirements.txt --unzip-setuptools \
           --setuptools bridgedb
 
   From now on, to use BridgeDB's virtualenv, just do ```$ workon bridgedb```
   (after sourcing virtualenvwrapper.sh, as before). To exit the virtualenv
   without exiting the shell, do ```$ deactivate```.

 - (virtualenv installs only) To install, set PYTHONPATH to include the root
   directory of the virtualenv:

        export PYTHONPATH=$PYTHONPATH:${VIRTUAL_ENV}/lib/python2.7/site-packages

 - Then proceed as usual:

        python setup.py install --record installed-files.txt

### Testing BridgeDB

To create a bunch of fake bridge descriptors to test BridgeDB, do:

        bridgedb mock [-n NUMBER_OF_DESCRIPTORS]

Then to run unittests, see the ```bridgedb test``` and ```bridgedb trial```
commands.

### Enabling additional features:

#### Translations

**Using New Translations**: This should be done when newly completed
  translations are available in Transifex.

Piece of cake. Running ```maint/get-completed-translations``` will take care
of cloning *only* the ```bridgedb_completed``` branch of Tor's
[translations repo](https://gitweb.torproject.org/translation.git) and placing
all the updated files in their correct locations.

**Requesting Translations for Altered/Added Source Code**: This should be done
  whenever any of the strings requiring translation -- _("they are formatted
  like this") -- are changed, or new ones are added.
  See ```lib/bridgedb/strings.py```.

Translations for Tor Project repos are kept
[in a separate repo](https://gitweb.torproject.org/translation.git).  You'll
need to extract the strings from BridgeDB's source code into .pot templates,
and place these .po files into the ```translation``` repo in the
```bridgedb``` branch. After than the .po files should be put into Transifex
(don't ask me how this works…) and translated. After the translations are
complete, the finished .po files should be placed into the
```bridgedb_completed``` branch.

 - To extract all strings from BridgeDB's source:

       python setup.py extract_messages --input-dirs ./lib/bridgedb/templates

   A .pot file will be created in ./i18n/templates/bridgedb.pot

 - Initialise catalogs for each desired language:

       python setup.py init_catalog -l LANG

   where ```LANG``` is the 2 or 4 letter country-code, eg. 'es'. If you've
   already initialised a particular language, do instead:
      
      python setup.py update_catalog


#### Enabling HTTPS
Create a self-signed certificate with:

     scripts/make-ssl-cert

Or, place an existing certificate in the path specified in bridgedb.conf by
the ```HTTPS_CERT_FILE``` option, and a private key where ```HTTPS_KEY_FILE```
points to. The defaults are 'cert' and 'privkey.pem', respectively.

#### CAPTCHAs

BridgeDB has two ways to use CAPTCHAs on webpages. The first uses reCaptcha_,
an external Google service (this requires an account with them), which
BridgeDB fetches the CAPTCHAs images from for each incoming request from a
client. The second method uses a local cache of pre-made CAPTCHAs, created by
scripting Gimp using gimp-captcha_. The latter cannot easily be run on
headless server, unfortunately, because Gimp requires an X server to be
installed.

.. _reCaptcha: https://www.google.com/recaptcha
.. gimp-capthca: https://github.com/isislovecruft/gimp-captcha

##### reCaptcha
To enable fetching CAPTCHAs from the reCaptcha API server, set these options
in bridgedb.conf:

    RECAPTCHA_ENABLED
    RECAPTCHA_PUB_KEY
    RECAPTCHA_SEC_KEY
    
##### gimp-captcha
To enable using a local cache of CAPTCHAs, set the following options::

    GIMP_CAPTCHA_ENABLED
    GIMP_CAPTCHA_DIR
    GIMP_CAPTCHA_HMAC_KEYFILE
    GIMP_CAPTCHA_RSA_KEYFILE

#### GnuPG email signing
Add these two options to your bridgedb.conf::

    EMAIL_GPG_SIGNING_ENABLED
    EMAIL_GPG_SIGNING_KEY
    
The former may be either True or False, and the latter must point to the
ascii-armored private key file. The keyfile must not be passphrase protected.

#### Preventing already-blocked bridges from being distributed
Uncomment or add ```COUNTRY_BLOCK_FILE``` to your bridgedb.conf. This file
should contain one bridge entry per line, in the format:
 
    fingerprint <bridge fingerprint> country-code <country code>

If the ```COUNTRY_BLOCK_FILE``` file is present, bridgedb will filter blocked
bridges from the responses it gives to clients requesting bridges.

#### Updating the SQL schema
Make sure that SQLite3 is installed. (You should have installed it already
during the setup and installation stage.) To update, do:

    sqlite3 path/to/bridgedist.db.sqlite

Enter the following commands at the ```sqlite>``` prompt: 

    CREATE TABLE BlockedBridges ( id INTEGER PRIMARY KEY NOT NULL, hex_key, blocking_country);
    CREATE INDEX BlockedBridgesBlockingCountry on BlockedBridges(hex_key);
    CREATE TABLE WarnedEmails ( email PRIMARY KEY NOT NULL, when_warned);
    CREATE INDEX WarnedEmailsWasWarned on WarnedEmails ( email );
    REPLACE INTO Config VALUES ( 'schema-version', 2 );
 

## Running BridgeDB

To run BridgeDB, simply make any necessary changes to bridgedb.conf, and
do: ```bridgedb```.

And remember that all files/directories in ```bridgedb.conf``` are assumed
relative to the runtime directory. By default, BridgeDB uses the current
working directory; you can, however specify an a different runtime directory::

    bridgedb -r /srv/bridges.torproject.org/run

Make sure that the files and directories referred to in bridgedb.conf
exist. However, many of them, if not found, will be touched on disk so that
attempts to read/write from/to them will not raise excessive errors.

When you have new lists of bridges from the Bridge Authority, replace the old
files and do::

    bridgedb --reload

Or just give it a SIGHUP::

    kill -s SIGHUP `cat .../run/bridgedb.pid`

#### To extract bucket files of all unallocated bridges:
Edit the configuration file value ```FILE_BUCKETS``` according to your
needs. For example, the following is a possible configuration:

    FILE_BUCKETS = { "name1": 10, "name2": 15, "foobar": 3 }

This configuration for buckets would result in 3 files being created for
bridge distribution: name1-2010-07-17.brdgs, name2-2010-07-17.brdgs and
foobar-2010-07-17.brdgs. The first file would contain 10 bridges from
BridgeDB's 'unallocated' pool. The second file would contain 15 bridges from
the same pool and the third one similarly 3 bridges. These files can then be
handed out to trusted parties via mail or fed to other distribution mechanisms
such as Twitter.

To dump all buckets to their files, send BridgeDB a ```SIGUSR1``` signal by
doing::

    kill -s SIGUSR1 `cat .../run/bridgedb.pid`

#### To use with HTTPS:
Just connect to the appropriate port.

#### To use with email: Any mail sent to the email port with a destination
username as defined by the EMAIL_USERNAME configuration option (default is
'bridge', e.g. bridges@...)  and sent from an ```@riseup.net```,
```@gmail.com```, or ```@yahoo.com``` address (by default, configured with the
EMAIL_DOMAINS option).

### Support
Send your questions to isis (A) torproject (circle) org.


CHANGELOG
Changes in version 0.6.3 - 2018-01-23

        * FIXES #24432 https://bugs.torproject.org/24432
        The production moat server had issues related to redirecting to
        resources properly, which are now fixed.

        * FIXES #24701 https://bugs.torproject.org/24701
        Adds a special surprise for the special someone who has been
        automatedly requesting bridges not through driving a browser, but
        through a script which is so thoroughly stupid that it doesn't
        even send the URL parameters for the CAPTCHA challenge and solution.
        Their script will now be delayed for quite some time and then
        rickrolled.  Mess with the best, die like the rest.

        * FIXES #24704 https://bugs.torproject.org/24704
        Bridges returned to a single request are now filtered such that
        there will never be two bridges from the same IPv4 /16 or IPv6 /64.

And includes the following general changes:

        * ADDS unittests for the legacy code in bridgedb/Bridges.py,
        bringing the total test coverage above 90% for the first time.


Changes in version 0.6.2 - 2017-12-20

        * FIXES #24636 https://bugs.torproject.org/24636
        The moat API specification included an extra response type which
        could be sent if there was no overlap between transports the
        client supported and those which the server supported.  This has
        been removed from the specification, which now describes the
        behaviour moat has always exhibited: if there is no overlap, the
        server responds with a CAPTCHA image response which includes the
        list of transports it does support.

        * FIXES #24637 https://bugs.torproject.org/24637
        The moat server did not respond correctly with the specified JSON
        API error type when there were no bridges available.  It now
        responds correctly with a 404 error whose details describe why the
        request could not be fulfilled.  The moat server also now logs
        messages if there were not the configured
        MOAT_BRIDGES_PER_RESPONSE number of bridges available.


Changes in version 0.6.1 - 2017-12-13

        * ADDS a shell script, scripts/test-moat, for testing either a
        locally-running moat server, or a remote one through a meek
        tunnel.
        Thanks to David Fifield for his work on meek, assistance setting
        it up, and providing the first version of this script.

        * FIXES #24433 https://bugs.torproject.org/24433
        The test-moat script wasn't sending an X-Forwarded-For header,
        which triggered a bug in the moat server, since the CAPTCHA
        solution includes an HMAC based on the client's IP (forwarded
        through all the several layers of tunnels/proxies).

        * FIXES #24443 https://bugs.torproject.org/24443
        Due to a difference between how booleans are parsed by Python's
        json library and normal Python booleans, the moat server was
        generating and returning QRCodes… regardless of whether the remote
        client application asked for one.  This is now fixed.

        * FIXES #24460 https://bugs.torproject.org/24460
        There was an unhandled error when sending certain (what appears to
        be possibly malicious? but in a very strange way) requests to
        BridgeDB's HTTPS distributor.  The robots making the requests were
        attempting to request bridges, but were presenting a CAPTCHA
        solution without the correct HTML form field parameters present,
        which isn't possible through normal usage of the web interface.
        Whoever or whatever is doing this is now going to be endlessly
        redirected so that they may forever spiral in their own private
        internet hell.  I reserve the right come up with a worse fate for
        them later, should I get bored.

        * FIXES #3015 https://bugs.torproject.org/3015
        BridgeDB has had a partially-implemented concept of "buckets"
        since the age of the dinosaurs: write some of the unallocated
        bridges to a file which should (somehow) be manually distributed.
        In addition to be unused and untested, there were several issues
        with the buckets, the most significant of which were the inability
        to request pluggable transports in a bucket and the fact that
        buckets were not persistent in any way (e.g. if i request a bucket
        of 50 bridges for Gomez and another with 50 for Morticia, they
        might end up with some of the same bridges, further, tomorrow
        they'll end up with 50 possibly different bridges than those they
        received today).  All of this code is now removed.

And includes the following general changes:

        * FIXES issues with JSON quote syntax and a mistaken JSON API
        "type" parameter in the specification of the moat server (in the
        README).
        Thanks to Mark Smith and Kathy Brade for pointing out the issues.


Changes in version 0.6.0 - 2017-11-15

        * ADDS a new JSON API distributor called "moat", which is intended
        for use for Tor Launcher to use to build an in-browser UI for
        retrieving bridges.

        * CHANGES the organisation of code to add a new
        bridgedb.distributors package as well as a
        bridgedb.distributors.common package for code shared between
        multiple distributors.


Changes in version 0.5.0 - 2017-10-28

        * FIXES #23957 https://bugs.torproject.org/23957
        BridgeDB now supports bridge operators choosing how their bridge
        will be distributed.  See the "BridgeDistribution" torrc option in
        tor's manpage for details.

        * FIXES #16650 https://bugs.torproject.org/16650
        BridgeDB is now accessible via select remote user interfaces
        through a meek tunnel.

        * FIXES #22998 https://bugs.torproject.org/23033
        * FIXES #23033 https://bugs.torproject.org/23033

bridgedb/Bucket.py
# -*- coding: utf-8 -*-

"""
This module is responsible for everything concerning file bucket bridge 
distribution. File bucket bridge distribution means that unallocated bridges 
are allocated to a certain pseudo-distributor and later written to a file.

For example, the following is a dict of pseudo-distributors (also called 
'bucket identifiers') with numbers of bridges assigned to them:

        FILE_BUCKETS = { "name1": 10, "name2": 15, "foobar": 3 }

This configuration for buckets would result in 3 files being created for bridge
distribution: name1-2010-07-17.brdgs, name2-2010-07-17.brdgs and 
foobar-2010-07-17.brdgs. The first file would contain 10 bridges from BridgeDB's
'unallocated' pool. The second file would contain 15 bridges from the same pool
and the third one similarly 3 bridges. These files can then be handed out to 
trusted parties via mail or fed to other distribution mechanisms such as 
twitter.

Note that in BridgeDB slang, the _distributor_ would still be 'unallocated',
even though in the database, there would now by 'name1', 'name2' or 'foobar'
instead of 'unallocated'. This is why they are called pseudo-distributors.
"""

import logging
import time
import bridgedb.Storage
import bridgedb.Bridges 
import binascii
import sqlite3
from gettext import gettext as _
toHex = binascii.b2a_hex


# What should pseudo distributors be prefixed with in the database so we can
# distinguish them from real distributors?
PSEUDO_DISTRI_PREFIX = "pseudo_"

# Set to rediculously high number
BUCKET_MAX_BRIDGES = 1000000


class BucketData(object):
    """Configures a bridge bucket with the number of bridges which should be
    allocated, the name of the bucket, and other similar data.

    :param str name: The name of this bucket (from the config file). This will
        be prefixed by the :data:`PSEUDO_DISTRIBUTOR_PREFIX`.
    :type needed: str or int
    :param needed: The number of bridges needed for this bucket (also from the
        config file).
    :param int allocated: Number of bridges already allocated for this bucket.
    """
    def __init__(self, name, needed):
        self.name = name
        if needed == "*":
            needed = BUCKET_MAX_BRIDGES
        self.needed = int(needed)
        self.allocated = 0

class BucketManager(object):
    """BucketManager reads a number of file bucket identifiers from the config.

    They're expected to be in the following format::

        FILE_BUCKETS = { "name1": 10, "name2": 15, "foobar": 3 }

    This syntax means that certain buckets ("name1", "name2" and so on) are
    given a number of bridges (10, 15 and so on). Names can be anything.  The
    name will later be the prefix of the file that is written with the
    assigned number of bridges in it. Instead of a number, a wildcard item
    ("*") is allowed, too. This means that the corresponsing bucket file will
    get the maximum number of possible bridges (as many as are left in the
    unallocated bucket).

    The files will be written in ip:port format, one bridge per line.

    The way this works internally is as follows:

    First of all, the assignBridgesToBuckets() routine runs through the
    database of bridges and looks up the 'distributor' field of each
    bridge. Unallocated bridges are sent to a pool for later assignement.
    Already allocated bridges for file bucket distribution are sorted and
    checked.  They're checked for whether their bucket identifier still exists
    in the current config and also whether the number of assigned bridges is
    still valid. If either the bucket identifier is not existing anymore or
    too many bridges are currently assigned to it, bridges will go to the
    unassigned pool.

    In the second step, after bridges are sorted and the unassigned pool is
    ready, the assignBridgesToBuckets() routine assigns one bridge from the
    unassigned pool to a known bucket identifier at a time until it either
    runs out of bridges in the unallocated pool or the number of needed
    bridges for that bucket is reached.

    When all bridges are assigned in this way, they can then be dumped into
    files by calling the dumpBridges() routine.

    :type cfg: :class:`bridgedb.persistent.Conf`
    :ivar cfg: The central configuration instance.
    :ivar list bucketList: A list of BucketData instances, holding all
        configured (and thus requested) buckets with their respective numbers.
    :ivar list unallocatedList: Holds all bridges from the 'unallocated' pool.
    :ivar bool unallocated_available: Is at least one unallocated bridge
        available?
    :ivar str distributor_prefix: The 'distributor' field in the database will
        hold the name of our pseudo-distributor, prefixed by this string. By
        default, this uses :data:`PSEUDO_DISTRIBUTOR_PREFIX`.
    :ivar db: The bridge database instance.
    """

    def __init__(self, cfg):
        """Create a ``BucketManager``.

        :type cfg: :class:`bridgedb.persistent.Conf`
        :param cfg: The central configuration instance.
        """
        self.cfg = cfg
        self.bucketList = []
        self.unallocatedList = []
        self.unallocated_available = False
        self.distributor_prefix = PSEUDO_DISTRI_PREFIX

    def addToUnallocatedList(self, hex_key):
        """Add a bridge by **hex_key** into the unallocated pool."""
        with bridgedb.Storage.getDB() as db:
            try:
                db.updateDistributorForHexKey("unallocated", hex_key)
            except:
                db.rollback()
                raise
            else:
                db.commit()
        self.unallocatedList.append(hex_key)
        self.unallocated_available = True

    def getBucketByIdent(self, bucketIdent):
        """If we know this bucket identifier, then return the corresponding
        :class:`BucketData` object.
        """
        for d in self.bucketList:
            if d.name == bucketIdent:
                return d
        return None

    def assignUnallocatedBridge(self, bucket):
        """Assign an unallocated bridge to a certain **bucket**."""
        hex_key = self.unallocatedList.pop()
        # Mark pseudo-allocators in the database as such
        allocator_name = bucket.name
        #print "KEY: %d NAME: %s" % (hex_key, allocator_name)
        logging.debug("Moving %s to %s" % (hex_key, allocator_name))
        with bridgedb.Storage.getDB() as db:
            try:
                db.updateDistributorForHexKey(allocator_name, hex_key)
            except:
                db.rollback()
                logging.warn("Failed to move %s to new distributor (%s)"
                             % (hex_key, allocator_name))

                # Ok, this seems useless, but for consistancy's sake, we'll
                # re-assign the bridge from this missed db update attempt to the
                # unallocated list. Remember? We pop()'d it before.
                self.addToUnallocatedList(hex_key)
                raise
            else:
                db.commit()
        bucket.allocated += 1
        if len(self.unallocatedList) < 1:
            self.unallocated_available = False
        return True

    def assignBridgesToBuckets(self):
        """Read file bucket identifiers from the configuration, sort them, and
        write necessary changes to the database.
        """
        logging.debug("Assigning bridges to buckets for pseudo-distributors")
        # Build distributor list
        for k, v in self.cfg.FILE_BUCKETS.items():
            prefixed_key = self.distributor_prefix + k
            d = BucketData(prefixed_key, v)
            self.bucketList.append(d)

        # Loop through all bridges and sort out distributors
        with bridgedb.Storage.getDB() as db:
            allBridges = db.getAllBridges()
        for bridge in allBridges:
            if bridge.distributor == "unallocated":
                self.addToUnallocatedList(bridge.hex_key)
                continue

            # Filter non-pseudo distributors (like 'https' and 'email') early,
            # too
            if not bridge.distributor.startswith(self.distributor_prefix):
                continue

            # Return the bucket in case we know it already
            d = self.getBucketByIdent(bridge.distributor)
            if d is not None:
                # Does this distributor need another bridge? If not, re-inject
                # it into the 'unallocated' pool for for later assignment
                if d.allocated < d.needed:
                    d.allocated += 1
                else:
                    # Bucket has enough members already, free this one
                    self.addToUnallocatedList(bridge.hex_key)
            # We don't know it. Maybe an old entry. Free it.
            else:
                self.addToUnallocatedList(bridge.hex_key)

        # Loop through bucketList while we have and need unallocated
        # bridges, assign one bridge at a time
        while self.unallocated_available and len(self.bucketList) > 0:
            logging.debug("We have %d unallocated bridges and %d buckets to " \
                          "fill. Let's do it."
                          % (len(self.unallocatedList), len(self.bucketList)))
            for d in self.bucketList:
                if d.allocated < d.needed:
                    try:
                        if not self.assignUnallocatedBridge(d):
                            break
                    except sqlite3.DatabaseError as e:
                        dist = d.name.replace(self.distributor_prefix, "")
                        logging.warn("Couldn't assign unallocated bridge to " \
                                     "%s: %s" % (dist, e))
                else:
                    # When we have enough bridges, remove bucket identifier 
                    # from list
                    self.bucketList.remove(d)

    def dumpBridgesToFile(self, filename, bridges):
        """Dump a list of given **bridges** into **filename**."""
        logging.debug("Dumping bridge assignments to file: %r" % filename)
        # get the bridge histories and sort by Time On Same Address
        bridgeHistories = []
        with bridgedb.Storage.getDB() as db:
            for b in bridges:
                if self.cfg.COLLECT_TIMESTAMPS:
                    bh = db.getBridgeHistory(b.hex_key)
                    if bh: bridgeHistories.append(bh)
                    bridgeHistories.sort(lambda x,y: cmp(x.weightedFractionalUptime,
                                     y.weightedFractionalUptime))

            try:
                f = open(filename, 'w')
                if self.cfg.COLLECT_TIMESTAMPS:
                    for bh in bridgeHistories:
                        days = bh.tosa / long(60*60*24)
                        line = "%s:%s\t(%d days at this address)" %  \
                               (bh.ip, bh.port, days)
                        if str(bh.fingerprint) in blocklist.keys():
                            line = line + "\t(Might be blocked): (%s)" % \
                                   ",".join(blocklist[bh.fingerprint])
                        f.write(line + '\n')
                else:
                    for bridge in bridges:
                        line = "%s:%d %s" \
                               % (bridge.address, bridge.or_port, bridge.hex_key)
                        f.write(line + '\n')
                f.close()
            except IOError:
                print "I/O error: %s" % filename

    def dumpBridges(self):
        """Dump all known file distributors to files, sorted by distributor."""
        logging.info("Dumping all distributors to file.")
        with bridgedb.Storage.getDB() as db:
            allBridges = db.getAllBridges()
        bridgeDict = {}
        # Sort returned bridges by distributor
        for bridge in allBridges:
            dist = str(bridge.distributor)
            if dist in bridgeDict.keys():
                bridgeDict[dist].append(bridge)
            else:
                bridgeDict[dist] = [bridge]

        # Now dump to file(s)
        for k in bridgeDict.keys():
            dist = k
            if (dist.startswith(self.distributor_prefix)):
                # Subtract the pseudo distributor prefix
                dist = dist.replace(self.distributor_prefix, "")
            # Be safe. Replace all '/' in distributor names
            dist = dist.replace("/", "_")
            filename = dist + "-" + time.strftime("%Y-%m-%d") + ".brdgs"
            self.dumpBridgesToFile(filename, bridgeDict[k])
